test/remote/gateways/remote_simetrik_test.rb

```diff
@@ -13,51 +12,203 @@ def setup
+    assert_equal response.avs_result['code'], 'G'
+    assert_equal response.cvv_result['code'], 'P'
     assert response.test?
   end
 
   def test_failed_authorize
     response = @gateway.authorize(@amount, @credit_card, @authorize_capture_options_failed)
     assert_failure response
     assert_instance_of Response, response
-    # assert_equal response.message, 'Successful authorize'
-    assert_equal response.params['code'][0], 'R', 'Should expected code like Rxxx'
-    assert_equal response.avs_result.to_hash['code'], 'I'
-    assert_equal response.cvv_result.to_hash['code'], 'P'
+    assert_not_equal response.error_code, nil, 'Should expected error code not equal to nil'
+    assert_equal response.avs_result['code'], 'I'
+    assert_equal response.cvv_result['code'], 'P'
+    assert response.test?
+  end
+
+  def test_failed_authorize_by_invalid_card
+    response = @gateway.authorize(@amount, @credit_card_invalid, @authorize_capture_options_success)
+    assert_failure response
+    assert_instance_of Response, response
+    assert_not_equal response.error_code, nil, 'Should expected error code not equal to nil'
+    assert_equal response.avs_result['code'], 'G'
+    assert_equal response.cvv_result['code'], 'P'
     assert response.test?
   end
 
   def test_success_purchase
     response = @gateway.purchase(@amount, @credit_card, @authorize_capture_options_success)
     assert_success response
     assert_instance_of Response, response
-    assert_equal response.params['message'], 'successful charge'
-    assert_equal response.params['code'][0], 'S', 'Should expected code like Sxxx'
-    assert_equal response.avs_result.to_hash['code'], 'G'
-    assert_equal response.cvv_result.to_hash['code'], 'P'
+    assert_equal response.message, 'successful charge'
+    assert_equal response.error_code, nil, 'Should expected error code equal to nil'
+    assert_equal response.avs_result['code'], 'G'
+    assert_equal response.cvv_result['code'], 'P'
     assert response.test?
   end
 
   def test_failed_purchase
     response = @gateway.purchase(@amount, @credit_card, @authorize_capture_options_failed)
     assert_failure response
     assert_instance_of Response, response
-    assert_equal response.params['code'][0], 'R', 'Should expected code like Rxxx'
-    assert_equal response.avs_result.to_hash['code'], 'I'
-    assert_equal response.cvv_result.to_hash['code'], 'P'
+    assert_not_equal response.error_code, nil, 'Should expected error code not equal to nil'
+    assert_equal response.avs_result['code'], 'I'
+    assert_equal response.cvv_result['code'], 'P'
+    assert response.test?
+  end
+
+  def test_failed_purchase_by_invalid_card
+    response = @gateway.purchase(@amount, @credit_card_invalid, @authorize_capture_options_success)
+    assert_failure response
+    assert_instance_of Response, response
+    assert_not_equal response.error_code, nil, 'Should expected error code not equal to nil'
```

Comment:
Thank you for adding this test! When I inspect the response, the error_code is `config_error`. Wouldn't we expect the gateway to return something like `invalid_card` here?

Reply:
Thank you for letting us know, we had a problem with our gateway and the test was receiving an unexpected error code. We have fixed this and, following your previous recommendation, opted for a narrower assert statement `assert_equal response.error_code, 'invalid_number'`.